### DISCUSSION TITLE
Get connected users on agent side

### DISCUSSION THREAD

**Discussion by trueg** [Ylianst/MeshCentral#6221] [Q&A] [answered]
AFAICT both meshcmd and meshctrl are intended to perform admin tasks on the mesh server side. Is there a way to request which users are currently connected on the agent side?
So in other words: on the same system which runs the MeshAgent I would like a script to print the names of the users currently connected.

**Comment by si458** [comment, marked as answer]
You can use the meshcentral assistant for this.
In the Web ui add agent then select the assistant, then set the type to monitor service, download the exe, then run the exe on the remote device, it will run in direct mode, but right click the tray icon and switch the service to ur serviceagent, then in future when a user runs the exe or clicks the tray icon, they can request help OR it will show if a user is connected or not :)

**Reply by trueg** [reply]
Thank you, that works. Now I need to figure out how to script that. :)

**Reply by si458** [reply]
the is/was a discussion/issue where i made with post on saying how u edit the registry to change the service from direct to your service
will post link/edit post when i find it!

**Reply by trueg** [reply]
oh, thanks. But what I actually meant was to get the information about connected users via a script rather than a user interface.

**Reply by si458** [reply]
@trueg ah right ok.
The only way u could do that is watch the browser console and the websocket data, then see which command returns what users are connected to devices, then open a websoclet to ur meshcentral server, login, then run that command u found!

**Comment by trueg** [comment]
For anyone who might be interested in something similar. I dug through the code of MeshCentralAssistant and came up with this little python class which does exactly what I need. Keep in mind that the assistant has a few ways to get to the node id but most of them involve looking at similar places in the windows registry.

```python
import base64
import json
import winreg


class MeshClient:
    """
    The MeshClient connects to the local MeshAgent on a Windows client.

    The connection is established locally through a named pipe. This connection is
    originally used by the MeshClientAssistant and not documented anywhere.

    The name of the pipe is stored in the Windows registry. While the original
    MeshAssistant code tried multiple way to get the node's id, we only use the
    one here which seems to be used for the test bench setup.

    Usage is very simple:

    ```
    with MeshClient() as mc:
        print(mc.connected_users())
    ```
    """

    def __init__(self) -> None:
        self.pipe = None

    def __enter__(self):
        # The following magic has been re-engineered from https://github.com/Ylianst/MeshCentralAssistant/blob/master/src/MeshAgent.cs
        reg = winreg.ConnectRegistry(None, winreg.HKEY_LOCAL_MACHINE)
        reg_key = winreg.OpenKey(reg, r"SOFTWARE\Open Source\Mesh Agent")
        node_id = winreg.QueryValueEx(reg_key, "NodeId")[0]
        reg_key.Close()
        reg.Close()

        pipe_id = (
            base64.b64decode(node_id.replace("@", "+").replace("$", "/")).hex()
            + "-DAIPC"
        )
        pipe_name = r"\\.\pipe" + "\\" + pipe_id

        self.pipe = open(pipe_name, "r+b", 0)
        return self

    def __exit__(
        self,
        exc_type,
        exc_value,
        traceback,
    ) -> None:
        self.pipe.close()

    def cmd(self, cmd: str, value: str = ""):
        """
        Send a command to the MeshAgent and retrieve the result.

        The protocol is very simple:
        - First the length of the data + 4 is sent as a 4 byte int
        - Then the command itself is sent which is a simple json structure

        The reply has a similar (if slightly different) structure:
        - First read 4 bytes to get the total length of the reply.
        - Then read the rest of the data based on the length in the first 4 bytes.
        """
        data = json.dumps({"cmd": cmd, "value": value}).encode("utf-8")
        self.pipe.write((len(data) + 4).to_bytes(4, byteorder="little"))
        self.pipe.write(data)
        r_len = int.from_bytes(self.pipe.read(4), byteorder="little")
        r_data = self.pipe.read(r_len).decode("utf-8")
        return json.loads(r_data)

    def connected_users(self):
        """
        Get the usernames of the users that are currently connected via a desktop session.
        """
        data = self.cmd("sessions")
        if "sessions" in data and "desktop" in data["sessions"]:
            desktop_users = list(data["sessions"]["desktop"].keys())
            return [user.split("//")[1] for user in desktop_users]
        return []


if __name__ == "__main__":
    with MeshClient() as mc:
        print(mc.connected_users())
```
